### DISCUSSION TITLE
[question] what is the purpose of ViewManagerDelegate

### DISCUSSION THREAD

**Discussion by vonovak** [reactwg/react-native-new-architecture#141] [Q&A] [answered]
When migrating a component to Fabric, there is a step documented here https://reactnative.dev/docs/next/the-new-architecture/pillars-fabric-components#centeredtextmanagerjava which adds the following

```
    private final ViewManagerDelegate<CenteredText> mDelegate;

    public CenteredTextManager(ReactApplicationContext context) {
        mDelegate = new RTNCenteredTextManagerDelegate<>(this);
    }

    @Nullable
    @Override
    protected ViewManagerDelegate<CenteredText> getDelegate() {
        return mDelegate;
    }
```

I'd like to know what the purpose of the delegate is. Is it mandatory?

Implementing the `RTNCenteredTextManagerInterface` does not enforce the presence of  `getDelegate()` and I'm wondering what it's purpose is, because all the delegate does is that is ends up calling the methods on the `ViewManager`. Is it for imperative command dispatching?

Thank you! :)

**Comment by cortinico** [comment, marked as answer]
Hi @vonovak,
Happy to respond,

> Implementing the RTNCenteredTextManagerInterface does not enforce the presence of getDelegate()

You're right in saying that the generated `Interface` does not require you to pass a delegate.
On the other hand, your `RTNCenteredTextManager` needs to extend one of the `ViewManager` classes (i.e. `SimpleViewManager` or `BaseViewManager` or just `ViewManager`).

The `ViewManager` class however has a `getDelegate()` method which can be used to pass a `ViewManagerDelegate`. This class is conveniently generated for you by the codegen, and will take care of all the wiring for the `setProperty` and `receiveCommand`. 

So in theory, you could avoid setting it, but your view won't work unless you wire all the `setProperty` manually. Plus the `BaseViewManagerDelegate` class which we extend from when we generate your delegate, has all the logic to handle common view props such as accessibility props and so on.
